NASA-AMMOS/aerie
# [AERIE-1967] Global Scheduling Conditions

* **Tickets addressed:** AERIE-1967
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description


Use case: Users want to help the scheduler avoid violating constraints. Currently, the scheduler has no knowledge of constraints (in the merlin sense of "constraints").

While it would be straightforward to give the scheduler access to merlin's constraints, we argue this would not provide much benefit with respect to the use case above. While the scheduler could _check_ that, after placing an activity, it has violated a constraint, it cannot _predict_ which regions of the plan would be safe for placing a given activity. We believe that a "guess and check" approach would be intractable - the scheduler needs some more help from the user.

This "help" that the user can provide the scheduler comes in the form of a "Global Scheduling Condition". This is a piece of typescript code, akin to a goal. Like a goal, it can be added to a scheduling specification, enabled, and disabled. Unlike a goal, it returns a `Windows` object, which uses the Constraints language to establish windows in which the scheduler is allowed to place activities. Also, unlike goals, Global Scheduling Conditions have no notion of priority.

## Verification

This PR has been validated through a rather tedious manual test. If I can get someone to pair with me, I'd love to turn this into an e2e test.
<details>
<summary>Test procedure</summary>

1. Clean build
2. Upload banananation
3. Check mission model id, you'll need it later
4. Create a new plan
5. Check the scheduling spec id, you'll need it later.
6. Create a new Goal
```typescript
export default (): Goal => {
  return Goal.ActivityRecurrenceGoal({
    activityTemplate: ActivityTemplates.BiteBanana({biteSize: 1}),
    interval: 24 * 60 * 60 * 1000 * 1000
  })
}
```
7. Run the following (fill in your model_id):
```graphql
mutation InsertGSC {
  insert_scheduling_condition_one(object:{
    model_id: 1
    name: "My first scheduling condition"
    definition: "export default (): Windows => Real.Resource('/fruit').greaterThan(5.0)"
  }) {
    id
  }
}
```
Note the id returned.
8. Run the following (fill in your spec id and condition id)
```graphql
insert_scheduling_specification_conditions_one(object:{
  condition_id: 1
  specification_id: 1
  enabled: true
}) {
  __typename
}
```
9. Run scheduling, check that it has not placed any activities.
10. Run the following to change the threshold from 5 to 1:
```graphql
mutation UpdateGSC {
  update_scheduling_condition_by_pk(
    pk_columns: {id:1},
    _set: {
      definition: "export default (): Windows => Real.Resource('/fruit').greaterThan(0.0)"
    }
  ) {
    __typename
  }
}
```
11. Run scheduling, check that it has placed activities. (Note that it should stop placing activities when the /fruit resource reaches 0.0)
</details>

## Documentation

- [x] Docs have not been written, but absolutely should be before merging this PR.

UPDATE: I have added a "Global Scheduling Conditions" section to the [scheduling guide](https://github.com/NASA-AMMOS/aerie/wiki/Scheduling-Guide)

## Future work

- This PR lays some groundwork for associating scheduling conditions with particular activity types (e.g. do not place "battery intensive" activities when the battery is low, but it's okay to place a "recharge" activity)
